std/internal/test/dummyrange.d

```diff
@@ -271,4 +307,251 @@ unittest
+        T it;
+        Cmp.init(it.arr);
+        for (size_t numRuns = 0; numRuns < 2; ++numRuns)
+        {
+            if (numRuns == 1)
+            {
+                static if (is(Unqual!(ElementType!(T)) == uint))
+                {
+                    it.reinit();
+                }
+
+                Cmp.init(it.arr);
+            }
+
+            assert(equal!(Cmp.cmp)(it, Cmp.iota(1, 11)));
+
+            static if (hasLength!T)
+            {
+                assert(it.length == 10);
+            }
+
+            assert(!Cmp.cmp(it.front, Cmp.dummyValue));
+            auto s = it.front;
+            it.front = Cmp.dummyValue;
+            assert(Cmp.cmp(it.front, Cmp.dummyValue));
+            it.front = s;
+
+            auto cmp = Cmp.iota(1,11);
+
+            size_t jdx = 0;
+            while (!it.empty && !cmp.empty)
+            {
+                static if (hasLength!T)
+                {
+                    assert(it.length == 10 - jdx);
+                }
+
+                assert(Cmp.cmp(it.front, cmp.front));
+                it.popFront();
+                cmp.popFront();
+
+                ++jdx;
+            }
+
+            assert(it.empty);
+            assert(cmp.empty);
+        }
+
+    }
+
+    static void testForwardRange(T,Cmp)()
+    {
+        T it;
+        Cmp.init(it.arr);
+        auto s = it.save();
+        s.popFront();
+        assert(!Cmp.cmp(s.front, it.front));
+    }
+
+    static void testBidirectionalRange(T,Cmp)()
+    {
+        T it;
+        Cmp.init(it.arr);
+        assert(equal!(Cmp.cmp)(it.retro, Cmp.iota().retro));
+
+        auto s = it.back;
+        assert(!Cmp.cmp(s, Cmp.dummyValue));
+        it.back = Cmp.dummyValue;
+        assert( Cmp.cmp(it.back, Cmp.dummyValue));
+        it.back = s;
+    }
+
+    static void testRandomAccessRange(T,Cmp)()
+    {
+        T it;
+        Cmp.init(it.arr);
+        size_t idx = 0;
+        foreach (jt; it)
+        {
+            assert(it[idx] == jt);
+
+            T copy = it[idx .. $];
+            auto cmp = Cmp.iota(idx + 1, it.length + 1);
+            assert(equal!(Cmp.cmp)(copy, cmp));
+
+            ++idx;
+        }
+
+        {
+            auto copy = it;
+            copy.arr = it.arr.dup;
+            for (size_t i = 0; i < copy.length; ++i)
+            {
+                copy[i] = Cmp.dummyValue;
+                copy[i] += Cmp.dummyValue;
+            }
+            assert(equal!(Cmp.cmp)(copy, Cmp.dummyValueRslt.repeat(copy.length)));
+        }
+
+        static if (it.r == ReturnBy.Reference)
+        {
+            T copy;
+            copy.arr = it.arr.dup;
+            for (size_t i = 0; i < copy.length; ++i)
+            {
+                copy[i] = Cmp.dummyValue;
+                copy[i] += Cmp.dummyValue;
+            }
+
+            assert(equal!(Cmp.cmp)(copy, Cmp.dummyValueRslt.repeat(copy.length)));
+        }
+    }
+
+    import std.meta : AliasSeq;
+
+    foreach (S; AliasSeq!(uint,double,TestFoo))
```

Comment:
More style nitpicks: please put a space after `,`, here and also elsewhere in this PR. It makes the result easier to read.